### PR TITLE
Creation of Formation Bundle

### PR DESCRIPTION
**Pull request by BartolottiLuca**
At the moment we support only mysql and postgresql helm releases and we don't add any database environment variable.

### CODE REVIEW

File: bundle/bundle-writer.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+						services []*common.Service,
+						databases []common.Database) error {
+
+	if templateRepository == "" {
+		//no stencil template defined for this pack, print an error and do nothing
+		fmt.Printf("Sorry but there is no stencil template for this language/framework yet\n")
+		return nil
+	}
+	//Create .bundle directory structure if it doesn't exist
+	skycapFolder := filepath.Join(outputDir, "skycap")
```
Comment (khash) [inline bundle/bundle-writer.go]: this should be done using OS temp directory / temp file APIs: https://yourbasic.org/golang/temporary-file-directory/

File: bundle/bundle-writer.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
+package bundle
```
Comment (khash) [inline bundle/bundle-writer.go]: Please always use `gofmt` and `golint` all the time. You can configure GoLand or VSCode to run it automatically.

File: bundle/bundle-writer.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+		skycapFolder,
+		manifestFile,
+		githubURL)
+
+	if err != nil {
+		return err
+	}
+
+	manifestFile, err = addMetadatas(manifestFile)
+
```
Comment (khash) [inline bundle/bundle-writer.go]: unneeded spacing

File: bundle/bundle-writer.go [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+		services,
+		skycapFolder,
+		manifestFile,
+		githubURL)
+
+	if err != nil {
+		return err
+	}
+
+	manifestFile, err = addMetadatas(manifestFile)
```
Comment (khash) [inline bundle/bundle-writer.go]: "metadata" and "data" don't have plural

File: bundle/bundle-writer.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+		return err
+	}
+	return err
+}
+
+// downloading templates from github and putting them into homedir
+func getStencilTemplateFile(templateRepository string, skycapFolder string, filename string, branch string) (string, error) {
+
+	//Download templates.json file
+	manifest_path := templateRepository + filename // don't need to use filepath since it's a URL
```
Comment (khash) [inline bundle/bundle-writer.go]: always use `path.Join` to make sure your code works on all OSs

File: bundle/bundle-writer.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	return nil
+}
+
+func getRequiredStencils(templateRepository string,
+						branch string,
+						outputDir string,
+						services []*common.Service,
+						skycapFolder string,
+						manifestFile ManifestBundle,
+						githubURL string) 	(ManifestBundle, error){
```
Comment (khash) [inline bundle/bundle-writer.go]: return should be `(*ManifestBundler, error)` and you should be able to use `return nil, err` if there is an error.

File: bundle/bundle-writer.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+		return ManifestBundle{}, err
+	}
+
+	var manifestStencils = make([]*BundleStencil, 0)
+
+	for i, data := range templJson {
+		if i =="templates" {
+			for _, stencils := range data.([]interface{}) {
+				stencil := stencils.(map[string]interface{})
+				if stencil["min_usage"].(float64) > 0 {
```
Comment (khash) [inline bundle/bundle-writer.go]: this might be ok here if we have a validation pre-step somewhere else.

File: bundle/bundle-writer.go [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+	newTemplate.Branch = branch
+	newTemplate.Stencils = manifestStencils
+
+	manifestFile.BaseTemplates =  append(manifestFile.BaseTemplates, &newTemplate)
+
+	return manifestFile, nil
+}
+
+
+func loadManifest() (ManifestBundle, error) {
```
Comment (khash) [inline bundle/bundle-writer.go]: use 
```
manifest := &ManifestBundle { 
Version: "1",
Metadata: nil,
...
``` 

instead

Comment (khash) [inline bundle/bundle-writer.go]: Return (*ManifestBundle, error)` instead

File: bundle/bundle-writer.go [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+func saveManifest(skycapFolder string, content ManifestBundle) error {
+	out, err := json.MarshalIndent(content, "", "  ")
+	if err != nil {
+		return err
+	}
+	manifestPath := filepath.Join(skycapFolder, "manifest.json")
+	return ioutil.WriteFile(manifestPath, out, 0600)
+}
+
+func saveEnvVars(prefix string, envVars map[string]string , manifestFile ManifestBundle, skycapFolder string) (ManifestBundle, error) {
```
Comment (khash) [inline bundle/bundle-writer.go]: Same with the return values

File: bundle/bundle-writer.go [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+	filename := prefix+"-config"
+	vars_path := filepath.Join(filepath.Join(skycapFolder,"configurations"), prefix+"-config")
+	var fileOut string
+	for key, value := range envVars {
+		fileOut = fileOut+key+"="+value+"\n"
```
Comment (khash) [inline bundle/bundle-writer.go]: it's much more readable and easier to debug if you add each item as a string in a `[]string` and then join them at the end with a `strings.Join(foo[:],"\n")`

File: bundle/bundle-writer.go [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+func downloadAndAddStencil(context string,
+							stencil map[string]interface{},
+							manifestFile ManifestBundle,
+							skycapFolder string,
+							templateRepository string,
+							branch string,
+							manifestStencils []*BundleStencil)	 (ManifestBundle, []*BundleStencil, error) {
+	var filename = ""
```
Comment (khash) [inline bundle/bundle-writer.go]: this part needs a rewrite. it's very convoluted. can be made much better

File: bundle/bundle-writer.go [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+	tempStencil.Sequence = int(stencil["preferred_sequence"].(float64))
+
+	manifestStencils = append(manifestStencils, &tempStencil)
+
+	return manifestFile, manifestStencils, nil
+}
+
+func addMetadatas(manifestFile ManifestBundle) (ManifestBundle, error) {
+	var metadata Metadata
+	metadata.Annotations = []string{"Generated by Cloud66 starter"}
```
Comment (khash) [inline bundle/bundle-writer.go]: Cloud 66 always has a ` ` between Cloud and 66

File: bundle/bundle-writer.go [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+	tempStencil.Tags = []string{"starter"}
+	tempStencil.Sequence = int(stencil["preferred_sequence"].(float64))
+
+	manifestStencils = append(manifestStencils, &tempStencil)
+
+	return manifestFile, manifestStencils, nil
+}
+
+func addMetadatas(manifestFile ManifestBundle) (ManifestBundle, error) {
+	var metadata Metadata
```
Comment (khash) [inline bundle/bundle-writer.go]: Use golang initializers

File: bundle/bundle-writer.go [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+	return manifestFile, nil
+}
+
+func addDatabase(manifestFile ManifestBundle, databases []common.Database) (ManifestBundle, error) {
+	var helm_releases = make([]*BundleHelmRelease, 0)
+	var release BundleHelmRelease
+	for _, db := range databases {
+		switch db.Name {
+		case "mysql":
+			release.Name = db.Name
```
Comment (khash) [inline bundle/bundle-writer.go]: Ultimately these should be moved out of the codebase. For now, it's ok.

File: main.go [hunk 15]
```diff
@@ -98,19 +84,20 @@ func init() {
 	-g dockerfile: only the Dockerfile
 	-g docker-compose: only the docker-compose.yml + Dockerfile
 	-g service: only the service.yml + Dockerfile (cloud 66 specific)
-	-g dockerfile,service,docker-compose (all files)
+	-g skycap: only the skycap files (cloud66 specific)
```
Comment (khash) [inline main.go]: Cloud 66

Comment (khash) [reply main.go]: There is no need to have this singled out, `skycap` is an option for `g` like the other ones.

File: main.go [hunk 16]
```diff
@@ -328,13 +264,14 @@ func analyze(
 
 	result := &analysisResult{Ok: false}
 
-	// if templateFolder is specified we're going to use that otherwise download
+	// if templateFolder is specified we're goi...........ng to use that otherwise download
```
Comment (khash) [inline main.go]: ??

File: packs/compose-to-service-yml/pack.go [hunk 17]
```diff
@@ -12,6 +12,12 @@ type Pack struct {
 	Analysis *Analysis
 }
 
+const (
+	StencilTemplatePath = "" //still not implemented
```
Comment (khash) [inline packs/compose-to-service-yml/pack.go]: This should have a `TODO` in the comments

File: packs/node/pack.go [hunk 18]
```diff
@@ -8,6 +8,11 @@ type Pack struct {
 	Analysis *Analysis
 }
 
+const (
+	nodeExpressStencilTemplatePath = "" // this way we only have to add the filename. We should start by download the templates.json, do a couples of checks and after that download the stuff
+	templateRepositoryBranch = "master"
```
Comment (khash) [inline packs/node/pack.go]: To enable testing, let this be a parameter into the CLI

File: bundle/bundle-writer.go (hunk 1 above)
Comment (BartolottiLuca) [reply bundle/bundle-writer.go]: I should create only the bundle base folder with the temp (this way will be unique) or all of them?

File: bundle/bundle-writer.go (hunk 5 above)
Comment (BartolottiLuca) [reply bundle/bundle-writer.go]: In this case, manifest_path is the URL that will be used to download the file (filename) from the templateRepository(BTR), so we need the '/' in any case. I can rename the `manifest_path` variable for it to be more clear.

File: bundle/bundle-writer.go (hunk 7 above)
Comment (BartolottiLuca) [reply bundle/bundle-writer.go]: I will create the needed structs.

File: packs/node/pack.go (hunk 18 above)
Comment (BartolottiLuca) [reply packs/node/pack.go]: Do you mean add a -btr flag to pass it as a param to the starter command?

File: bundle/bundle-writer.go (hunk 1 above)
Comment (khash) [reply bundle/bundle-writer.go]: If you need to have a temporary folder anywhere in the system which will be removed after you're done, it should be in the Temp folder, using the golang OS temp folder functions.

**Pull request merged**